guppylang/definition/custom.py

```diff
@@ -1,129 +1,145 @@
+    ) -> list[OutPortV]:
+        """Compiles a call to the function."""
+        self.call_compiler._setup(type_args, dfg, graph, globals, node)
+        return self.call_compiler.compile(args)
+
+
+@dataclass(frozen=True)
+class CustomFunctionDef(RawCustomFunctionDef, CompiledCallableDef):
+    """A custom function with parsed and checked signature."""
 
-        try:
-            self._ty = check_signature(self.defined_at, globals)
-        except GuppyError:
-            # We can ignore the error if a custom call checker is provided and the
-            # function may not be used as a higher-order value
-            if self.call_checker is None or self.higher_order_value:
-                raise
+    ty: FunctionType
 
     def check_call(
         self, args: list[ast.expr], ty: Type, node: AstNode, ctx: Context
     ) -> tuple[ast.expr, Subst]:
+        """Checks the return type of a function call against a given type.
+
+        This is done by invoking the provided `CustomCallChecker`.
+        """
         self.call_checker._setup(ctx, node, self)
         new_node, subst = self.call_checker.check(args, ty)
         return with_type(ty, with_loc(node, new_node)), subst
 
     def synthesize_call(
         self, args: list[ast.expr], node: AstNode, ctx: "Context"
     ) -> tuple[ast.expr, Type]:
+        """Synthesizes the return type of a function call.
+
+        This is done by invoking the provided `CustomCallChecker`.
+        """
         self.call_checker._setup(ctx, node, self)
         new_node, ty = self.call_checker.synthesize(args)
         return with_type(ty, with_loc(node, new_node)), ty
 
-    def compile_call(
-        self,
-        args: list[OutPortV],
-        type_args: Inst,
-        dfg: DFContainer,
-        graph: Hugr,
-        globals: CompiledGlobals,
-        node: AstNode,
-    ) -> list[OutPortV]:
-        self.call_compiler._setup(type_args, dfg, graph, globals, node)
-        return self.call_compiler.compile(args)
-
     def load(
         self, dfg: "DFContainer", graph: Hugr, globals: CompiledGlobals, node: AstNode
     ) -> OutPortV:
         """Loads the custom function as a value into a local dataflow graph.
 
-        This will place a `FunctionDef` node into the Hugr module if one for this
-        function doesn't already exist and loads it into the DFG. This operation will
-        fail if no function type has been specified.
+        This will place a `FunctionDef` node into the Hugr module and loads it into the
+        DFG. This operation will fail the function is not allowed to be used as a
+        higher-order value.
         """
-        if self._ty is None:
+        # TODO: This should be raised during checking, not compilation!
+        if not self.higher_order_value:
             raise GuppyError(
                 "This function does not support usage in a higher-order context",
                 node,
             )
 
-        if self._ty.parametrized:
+        if self.ty.parametrized:
```

Comment:
I'm not sure where this `self.ty` is getting set

Reply:
It is set while parsing the `RawCustomFunctionDef`:

https://github.com/CQCL/guppylang/blob/05c5724962e6a5841ec2caf3f688ed9fbdad0808/guppylang/definition/custom.py#L69-L82